### ISSUE TITLE
REQ/Discussion RE: Custom LANG/Textures/Assets

### ISSUE THREAD

**Issue by kronflux** [fgsfdsfgs/perfect_dark#167]
As the game is now in a fully playable state where the issues from hereon will be relatively minor fixes and optimizations, I'm wondering what the first steps would be working toward supporting custom assets(ie: custom LANG/text, custom or higher resolution textures, sounds, models, etc.)
I recall that the original SM64 ports dumped the assets from the ROM and used them in the compilation process, and I'm assuming that's where we would need to start, so the question would be how to be load these assets rather than using the built-in assets from the ROM?

**Comment by fgsfdsfgs** [comment]
There is already a way to load external replacements for ROM assets, but it is limited.

For ROM asset files (i.e. stuff that's loaded by the `fileLoad` family of functions: e.g. models, levels, language banks), you can put replacements into the `data/files` directory, with the same names as they have in the ROM file table. For example, if you wanted to replace the english string table for the Villa mission, you'd put the replacement at `data/files/LeldE`. Note that it has to be in the same format as it is in the ROM (big endian, same structure), but it can be either zlib-compressed or not.

For entire ROM segments (e.g. fonts, mpstrings, texture data, sound data) you can put replacements into the `data/segs` directory, with the same names they have in the big segment table in `port/src/romdata.c`. Of course that means you can't actually provide replacements for separate textures, sequences and sound effects, nor is there support for HD textures. Not to mention the `textureconfigs` segment is not currently loaded from the ROM, meaning it would be difficult to add extra textures in some cases.

I have not tested this very thoroughly and there is currently no way to actually *add* extra files to the game. Also still need to figure out where to put custom strings. I've outlined [some thoughts](https://github.com/fgsfdsfgs/perfect_dark/issues/144#issuecomment-1704421442) about that in another issue.

**Comment by mmillar-bolis** [comment]
What is the correct method to obtain files like `LeldE`? Extracting assets under the normal `perfect_dark` decompilation project does not seem to produce those text blobs. Is the proper method to use the `gesetupeditor` toolkit that others have mentioned? Unfortunately, I can't seem to produce a file in the proper format. I've tried byte-swapping, reversing order, etc. but the game still crashes.

**Comment by fgsfdsfgs** [comment]
`make extract` does produce them though, they're in the `extracted/files` directory. You might have to rename them though. For example, files in the `lang` directory should just have the `L` prefix instead of being in a separate `lang` folder: `extracted/files/lang/optionsE.bin` -> `data/files/LoptionsE`.
However it does seem that just doing that absolutely murders the options text bank. Maybe it does actually have to be zlib-compressed.
UPD: Pushed a workaround for that. Now it loads.
UPD2: I am fucking stupid. This is not gonna be that easy. For now you'll have to match the format exactly, including the 1173 compression. Files can be compressed using `tools/rarezip`, though that only works on Linux out of the box.

**Comment by mmillar-bolis** [comment]
Thank you so much! It appears that I was running into problems with directory creation that were solved by `make allclean`, so I very much apologize for not having tried that first before re-extracting. You are absolutely correct that the blobs are extracted.

I was able to poke a change, recompress, and load the file without crashes. And yeah, I saw the blank objectives too, so I'll go ahead and repull and rebuild. Thank you again for your help with this!

**Comment by fgsfdsfgs** [comment]
Thanks for testing. Ideally it should be able to load both compressed and uncompressed resources, as well as maybe the "loose files" that GE Setup Editor produces.

**Comment by mmillar-bolis** [comment]
Absolutely! I went ahead and did some more testing. If I compress `eldE.bin` without any changes and load it under `files/LeldE`, all text appears correctly when selecting each difficulty for Villa.

However, if I modify the Briefing text for the Special Agent difficulty (changing the second instance of `Objective Two` to `Objective Three` as mentioned in #154 ) the file of course grows by two bytes. However, the modified text does successfully appear under the Briefing menu, as can be seen here:

![objectives_001](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/88b0ea0f-22ad-482d-8de7-bc22fa936115)

But as you mentioned, the Overview text is dropped:

![objectives_002](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/05ad92d0-5336-4947-83fd-6613a9aa2c47)

I also noticed that the Briefing text for the Agent difficulty is also dropped:

![objectives_003](https://github.com/fgsfdsfgs/perfect_dark/assets/62162171/f7c54d66-5a28-4e57-bf36-29ac57d30861)

Because the Agent Briefing section sits below the Special Agent Briefing section in `eldE.bin`, it would appear that anything below a section that is modified gets dropped.

I went ahead and fetched a clean copy of `eldE.bin` and added two bytes into the Perfect Agent Briefing section near the top of the file, then zipped it and loaded the game. Surely enough both the Agent and Special Agent Briefings were now blank.

**Comment by fgsfdsfgs** [comment]
This is because a lang file consists of two parts: an offset table that points to strings and the strings themselves. When you insert data in the middle of a string, it moves the other strings away from the location pointed to by their offsets, so you have to also edit those.
To avoid this you could edit the JSON lang files in the decomp (or in this repo) and then use `mklang` to generate a new lang bank, or maybe try using the GE Setup Editor, I think it has a lang edit fuction.

**Comment by mmillar-bolis** [comment]
Doh! Shows what I know, haha. I took your advice and went with `mklang`, which worked great! Though, I had to strip the elf header and the last four bytes, but it otherwise produced a fully working file. Plus I'm getting to know things a lot better, so thank you for all your help!
